### PR TITLE
Cancelable cats.effect.IO

### PR DESCRIPTION
**Pull request by alexandru**
If there's consensus, this is a proof of concept for a cancelable `cats.effect.IO`.

It is inspired by [Monix](https://github.com/monix/monix/)'s `Task`, using approximately the same cancellation mechanism, minus some features.

This doesn't necessarily have to be approved, however the implementation is solid and I'm pushing it because I've come to the conclusion that the type class hierarchy in `cats-effect` is effectively crippled due to `IO` not being cancelable.

I came to this conclusion after trying to describe this data type:

```scala
trait Timer[F[_]] {
  def async: Async[F]
  def shift: F[Unit]

  def sleep(timespan: FiniteDuration): F[Unit]
}
``` 

This is basically the pure equivalent of [ScheduledExecutorService](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html) and of JavaScript's [setTimeout](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout), an operation that's fundamental for *any platform* capable of asynchronous processing and that we need to describe in a pure way.

## Implementation notes

1. not all `IO` tasks are cancelable, only tasks built with `IO.cancelable` can cancel the evaluation (see below)
  - should go without saying (after point 1) that `flatMap` chains are not auto-cancelable, and compared with Monix's `Task` there's no way to configure an `IO` task to be auto-cancelable
  - if this is a problem, `flatMap` loops can be made cancelable by using `IO.cancelBoundary`, see below
2. `IO` tasks that are cancelable, become non-terminating on `cancel`
  - they can be turned into tasks that trigger an error with `onCancelRaiseError`, which can be used for implementing `bracket` or other operations, see below
3. this PR does NOT break API compatibility
4. this PR does NOT break semantics for current code, given that cancelable tasks can only be built with `IO.cancelable`, whereas tasks built with `IO.async`, `IO.apply` or other builders are not cancelable

### Versus the "async interruption" from Haskell and Scalaz 8

On "cancel" what happens is that whatever you specify in the `IO.cancelable` builder gets executed. And depending on the implementation of an `IO.cancelable` task, it can become non-terminating. And our impure `cancel()` is a `() => Unit`.

By comparison Haskell and Scalaz sends an error, a `Throwable` on interruption and cancelled tasks get completed with that `Throwable`. Their impure cancel is a `Throwable => Unit`.

We on the other hand have an `onCancelRaiseError(e: Throwable)`, which can transform a task that's non-terminating on cancel, into one that raises an error on cancel. This operation also creates a race condition - cutting off the signaling to downstream, even if the source is uncancelable.

`Throwable => Unit` allows the task's logic to know the cancellation reason, however I have yet to see a good use case for it — usually cancellation is about cutting the connection to the producer, closing all resources as soon as possible, because you're no longer interested in the result, due to some race condition that happened.

`Throwable => Unit` is also a little confusing, being too broad in scope. Users might be tricked into sending messages back to the producer via this channel, in order to steer it, to change its outcome - however cancellation is cancellation, we're doing it for the purpose of releasing resources and the implementation of race conditions will end up closing the connection, disallowing the cancelled task to send anything downstream.

Therefore it's confusing for the user and the only practical use is to release resources differently, based on the received error. But that's not a use-case I feel that's worth pursuing, given the increase in complexity.

## IO.cancelable builder

`IO.async` remains the same and we now have a new `IO.cancelable` builder - and in the provided function the user is expected to return a "cancelable", which is a `() => Unit` function that is supposed to cancel the started asynchronous computation. For example:

```scala
def sleep(duration: FiniteDuration)(implicit sc: ScheduledExecutorService): IO[Unit] =
  IO.cancelable { cb => 
    val c = sc.schedule(() => cb(Right()), duration.delay, duration.unit)
    // Returning a function that can cancel our IO evaluation
    () => c.cancel()
  }
```

## start + Fiber + cancel as a pure action

`IO` values that get started in the `IO` context (e.g. forked) get represented with a new interface:

```scala
trait Fiber[F[+_], +A] {

  def cancel: F[Unit]

  def join: F[A]
}
```

And `IO` now has a new `start` method on it exposing it:

```scala
sealed abstract class IO[+A] {

  def start: IO[Fiber[IO, A]]
}
```

A `Fiber` can be cancelled or joined, representing a background process that may run concurrently.
NOTE: we don't fork threads automatically, we just start the `IO` and if the `IO` happens to run asynchronously, then we have actual parallelism, otherwise we don't.

Example:

```scala
for {
  fiber <- IO.shift *> launchMissiles.start
  _  <- runToBunker.handleErrorWith { e =>
    // Retreating to bunker failed, abort!
    fiber.cancel *> IO.raiseError(e)
  }
  aftermath <- fiber.join
} yield {
  aftermath
}
```

## unsafeRunCancelable

```scala
// This is the sleep 
val io: IO[Unit] = sleep(10.seconds) *> IO.eval(println("Hello!")

val cancel: () => Unit =
  io.unsafeRunCancelable(r => println(s"Done: $r"))

// ... if a race condition happens, we can cancel it ...
cancel()
```

## uncancelable marker

Given a cancelable `IO`, we can turn it into an `IO` that cannot be cancelled:

```scala
// Our reference from above
val io: IO[Unit] = sleep(10.seconds) *> IO.eval(println("Hello!")

// This IO can't be canceled, even if we try
io.uncancelable
```

## onCancelRaiseError

Normally `IO` tasks that are cancelable, like our `sleep` above, become non-terminating on cancellation.

Given any `IO` (cancelable or not), the `onCancelRaiseError` operator creates a race condition between the source and cancellation ... and if cancellation happens, the cancel signal gets sent to the source and the resulting IO completes with the specified error.

```scala
import java.util.concurrent.CancellationException 

// Our reference from above
val io: IO[Unit] = sleep(10.seconds) *> IO.eval(println("Hello!")

// In case of cancellation, complete in error
val io2 = io.onCancelRaiseError(new CancellationException("Boo"))

val cancel: () => Unit =
  io.unsafeRunCancelable(r => println(s"Done: $r"))

// ... if a race condition happens, we can cancel it ...
cancel()
// Prints:
// Done: Left(CancellationException: Boo)
```

## IO.cancelBoundary

Returns a cancelable boundary — an `IO` task that checks for the cancellation status of the run-loop and does not allow for the bind continuation to keep executing in case cancellation happened.

This operation is very similar to `IO.shift`, as it can be dropped in `flatMap` chains in order to make loops cancelable.

```scala
def fib(n: Int, a: Long, b: Long): IO[Long] =
  IO.suspend {
    if (n <= 0) IO.pure(a) else {
      val next = fib(n - 1, b, a + b)

      // Every 100-th cycle, check cancellation status
      if (n % 100 == 0)
        IO.cancelBoundary *> next
      else
        next
    }
  }
```

## parMap2

This PR updates the `parMap2` implementation to cancel the other task if any of them fails — thus no longer waiting on both to complete in order to signal an error.

### CODE REVIEW

**Comment by edmundnoble** [comment]
Can you explain how having cancel as a pure function attached to an `IO[A]` works?

Here's an example of what I mean. Say I have an IO action here:
`def launchMissiles: IO[Unit]`

I run two copies of `launchMissiles` concurrently; I want to cancel one of the missile launches. I use `launchMissiles.cancel`; is it not the case that because `launchMissiles` here is a `def` and not a `val`, that doesn't actually cancel anything, because it's a new `IO`? Is it not also the case that if I used a `val`, *all* launches have to be cancelled?

It just seems wrong to me to have a cancellation handle, which is morally state that is generated by the starting of an `IO` action, be accessible *before* the `IO` action has actually begun.

Because there not being a difference between using a `val` and a `def` is actually one of the reasons why `IO` is important. Because if there *is* a difference then it is *not pure*. That is the exact definition of referential transparency.

**Comment by alexandru** [comment]
@edmundnoble the pure `cancel` action makes sense in the context of operations creating race conditions, which haven't been described in this PR - because I wanted to keep it focused on cancellation only. The test however describes an operation such as this:

```scala
  def fork[A](fa: IO[A])(implicit ec: ExecutionContext): IO[IO[A]] =
    IO.async { cb =>
      val p = Promise[A]()
      val conn = ForwardCancelable()
      // Executing `fa` on another thread
      ec.execute(new Runnable {
        def run(): Unit = {
          conn := fa.unsafeRunCancelable {
            case Right(a) => p.success(a)
            case Left(e) => p.failure(e)
          }
        }
      })
      // Signaling a new, memoized IO reference
      cb(Right(
        IO.cancelable { cb =>
          p.future.onComplete {
            case Success(a) => cb(Right(a))
            case Failure(e) => cb(Left(e))
          }(immediate)
          // Returning cancellation reference
          conn
        }))
    }
```

So now, we can create a race condition:

```scala
for {
  fa <- fork(launchMissiles)
  _  <- runToBunker.handleErrorWith { e =>
    // Retreating to bunker failed, abort!
    fa.cancel *> IO.raiseError(e)
  },
  result <- fa
} yield result
```

Note that in this case we are talking of an `IO[A]` reference attached to a running process.

**Comment by johnynek** [comment]
@alexandru I admire the work here. Twitter futures are cancelable (but not referentially transparent of course) and I have worked with them a fair amount.

That said I have never liked cancellation since it seems hard to reason about: we can’t be sure a cancelation takes (in for instance the example you give above with the missiles). How do you think about this? I know you have a lot of experience with Monix with this.

In an ideal world we could have atomic IO: the whole thing works or state doesn’t change. That is obviously very hard since you need distributed transactions in the general case. Maybe cancelation is no worse than general failure since failure leaves you in a state where you can’t be sure where exactly the IO failed.

Anyway, I’d be interested in your thoughts about cancelation in general and maybe some real world examples where you really want it. Is it about recovering computational resources? Stop working on things that won’t be useful?

**Comment by edmundnoble** [comment]
I don't believe that your comment answers any of my questions. I'll try to phrase them more clearly.

1. Are you or are you not telling me that functions that create IO actions are impure with this cancellation strategy, because they cannot be substituted with the IO actions themselves safely?

2. Is it not the case that `start` should actually provide an `IO[(IO[A], IO[Unit])]`? Why not?

3. How can we actually abstract over this `cancel` action; i.e., write a meaningful method that takes an `IO[A]` as a parameter, and calls cancel on it.

4. What happens with the edge cases I provided earlier: cancelling an IO action before running it, and rerunning an IO action after cancelling it. Perhaps even more importantly importantly, cancelling an IO action while running it multiple times concurrently.

5. Is there no way to make a cancellable `IO[A]` that isn't attached to a running process (that was produced inside IO itself)? Because if there isn't, you're essentially using the Scalaz 8 model of fiber cancellation but with much weaker types. Providing an example and usage of a cancellable `IO[A]` which is not generated in IO itself (i.e. `IO[IO[A]]`) would help clear this up. If that isn't practically useful, then there's no getting around the fact that the types here are more loose than they need to be.

You can totally provide a pure `cancel`, but I'm not convinced in the slightest that it makes sense as a member on `IO` and not `Fiber` (or whatever would be returned by `IO.fork`; hell, call it `CancellationHandle` if you want).

**Comment by alexandru** [comment]
@edmundnoble I understand your point, but for one It's easier to get answers by taking a look at the [implementation of cancel](https://github.com/typelevel/cats-effect/pull/121/files#diff-66b5ee7de7890355cdb1cc56350d76f4R29), as there's nothing special about it.

If I were to introduce a "Fiber" type, it would literally be this:

```scala
final case class Fiber[+A](val join: IO[A]) {
  def cancel: IO[Unit] = IOCancel.signal(join)
}
```

Now it might be more intuitive from a types perspective on usage, but from an implementation perspective it's just pretending something that isn't there, as this `cancel` is defined for `IO`. And in fact users can define this operation themselves:

```scala
def cancel[A](fa: IO[A]): IO[Unit] =
  IO.async { cb =>
    // Yes, from an implementation perspective, it's the equivalent of 
    // executing the source, followed by a cancel ...
    val cancel = fa.unsafeRunCancelable(cb)
    cancel() // triggering cancel
    cb(Right(()))
  }
```

> Are you or are you not telling me that functions that create IO actions are impure with this cancellation strategy

I'm telling you is that the operation becomes useful for `IO` actions attached to an already running process, like what happens with `fork` above. But its definition simply exist and given that it can be defined by users themselves, for as long as they get an `unsafeRunCancelable`, I didn't see the point in inventing a new type.

That said, the pure `cancel` is a red herring — I wouldn't mind adding a new `Forked` or `Fiber` type, plus at this point without an operation like `start`, `fork` or `racePair` it's basically a distraction that I could remove from the PR.

You drive a good argument though. If others feel the same way, then `Forked` / `Fiber` will happen.

**Comment by alexandru** [comment]
@johnynek I hear you. Will think of some samples where cancellation is useful, can't write another long message right now, just dropping a note that I'll compile something.

**Comment by LukaJCB** [comment]
FWIW, as a Person not super well versed with most of the internals of this repo, I think adding a `Forked` makes the API much more clear in how to use it. I understand the current formulation is more honest, but maybe some times it's better to be less honest and hide the implementation details.

**Comment by alexandru** [comment]
Very well then, I've now added the `Fiber` interface, along with the `start` method (otherwise there's no way to expose it).

/cc @edmundnoble @LukaJCB

**Comment by bkirwi** [comment]
I'm not sure I follow... if `flatMap` chains aren't automatically cancellable, doesn't that break the right identity law for Monad? 

```scala
cancellable.flatMap(IO.pure) != cancellable
```

**Comment by alexandru** [comment]
@bkirwi that sample isn’t true, no.

**Comment by alexandru** [comment]
@bkirwi to add more details - the implementation doesn't make `flatMap` chains to be auto-cancelable, but the cancellation signal does get propagated until a boundary that can do something with it, so applying `flatMap` on a cancellable `IO` will yield a cancellable `IO`.

Also see the sample for `IO.cancelBoundary`.

**Comment by mpilquist** [comment]
This looks great to me. We need to add cancelation to the type class hierarchy somewhere in order to allow middleware to take advantage of cancelation. Are you planning on adding that?

File: laws/shared/src/main/scala/cats/effect/laws/util/TestInstances.scala [hunk 1]
```diff
@@ -71,28 +68,20 @@ trait TestInstances {
...
             }
-          case Some(Failure(ex1)) =>
+          case Some(Failure(_)) =>
             y.value match {
-              case Some(Failure(ex2)) =>
-                extractEx(ex1) == extractEx(ex2)
+              case Some(Failure(_)) =>
+                // All exceptions are non-terminating and given exceptions
+                // aren't values (being mutable, they implement reference
+                // equality), then we can't really test them reliable,
```
Comment (rossabaker) [inline laws/shared/src/main/scala/cats/effect/laws/util/TestInstances.scala]: *reliably

**Comment by alexandru** [comment]
@mpilquist @rossabaker type classes are harder to developer, which is why I think `IO` needs to be modified first, or otherwise we'll choke on making choices 😀

Besides Monix's `Task` and this `IO`, a third implementation we can use as a sample is Scalaz 8 `IO` which is closer to Haskell's `IO` and I think we can take a look at what projects from other communities have been doing, like [purescript-fork](https://github.com/slamdata/purescript-fork/).

On the existing type classes, `Sync` is definitely still good. We'll need to reach a decision on `Async` and `Effect` — either they need to express cancellation, or we leave them as is and express new stuff in other type classes:

- for `Async` because we are left without an implementation like the current `IO`, it's now debatable on whether it's still needed; we could add `IO.cancelable` to it with a default implementation and thus keep binary compatibility for Scala 2.12
- for `Effect`, `runAsync` is still OK, but due to its return type and behavior (as it does not wait for the triggering of that callback), we can't cancel the running computation — so either we change the behavior (such that the returning `IO[Unit]` awaits completion), or we change the signature in `IO[IO[Unit]]`; or we introduce another type class

On new type classes, an important first step is @LukaJCB's work in #113 on adding a `Bracket` type class, which we've ensured that it works in the presence of cancellation as well.

The type class hierarchy also needs to express:

- the `IO.cancelable` builder
- `start` and `cancel`

So there's some work to do, but the first steps are to agree on:

1. accepting this new `IO` implementation
2. whether we can break compatibility in the current type class definitions, or if we restrict ourselves to adding new operations with default implementations, or new type classes, both of which are possible

**Comment by pchlupacek** [comment]
@alexandru any reason why cancel is not using `IO[Unit]` instead of `() => Unit` as its result type ?

**Comment by alexandru** [comment]
> any reason why cancel is not using IO[Unit] instead of () => Unit as its result type ?

That's a very good question @pchlupacek and one that I've often thought about. There would be two cases for which you'd want `IO[Unit]` in there:

1. Suspending side effects, but a simple, side-effectful `() => Unit` is enough and lets be honest, there's nothing pure that goes on in the function passed to `IO.cancelable`, being essentially FFI
2. Back-pressuring on a completion acknowledgement, which is wrong (explained below) and by using `IO[Unit]` you're signaling the wrong semantics (i.e. `() => Unit` signals a pretty clear behavior in that regard)

Back-pressuring on a cancellation acknowledgement is a pretty bad idea in general, as can be seen in all implementations that tried doing it. For example this was actually discussed in the [reactive-streams specification](https://github.com/reactive-streams/reactive-streams-jvm/issues/52).

When it comes to `IO[A]`, given its ability to model asynchronous computations, we are talking about an asynchronous producer/consumer communication protocol. And on cancellation you want to force close the connection and forget that the producer ever existed. Requiring an acknowledgement of cancellation from the producer is often not possible, especially because cancellation is often the result of a race condition that shouldn't happen in normal conditions (i.e. the producer is timing out).

I once worked on a web service participating in bidding exchanges with massive traffic and strict latency requirements. At some point due to integrating with a shitty bidding exchange that had a broken HTTP client, we ended up with the effects of [Slowloris](https://en.wikipedia.org/wiki/Slowloris_(computer_security)) because thousands of connections were left in either CLOSE_WAIT or TIME_WAIT, waiting for the kernel to time them out, with the servers reaching the maximum number of connections possible, with the whole cluster crashing at some point. And this was because our server was too fair, terminating TCP connections gracefully.

For `IO` in particular, it's very enlightening to see how how the pure `.cancel` works, [see code](https://github.com/typelevel/cats-effect/pull/121/files#diff-66b5ee7de7890355cdb1cc56350d76f4R29).

Basically you first execute the source, but ensure that it gets cancelled by passing a `StackedCancelable.alreadyCancelled` reference to the run-loop. What this should highlight is the nature of cancellation ... at some point, some logic in that run-loop will end up depending or noticing this `alreadyCancelled` reference and will act accordingly, but trying to back-pressure on that is a little pointless, because you don't have a guarantee of what such a signal would mean — i.e. was some connection closed, or was the run-loop terminated due to a `cancelBoundary` that noticed `isCanceled == true`?

So the only reason to want `IO[Unit]` is to back-pressure for its result, but there are many cases in which back-pressuring for that cancellation acknowledgement is a land mine. And otherwise it signals the wrong semantics.

Also another very reason is that it would complicate the implementation. This isn't a very good reason to be against something, but it does raise the acceptability threshold, as in, anything that complicates the implementation needs really good reasons.

**Comment by pchlupacek** [comment]
@alexandru not sure if I follow everything you mention. Essentially I think that any side-effecting code in library has to be protected by suspend // F // IO. The fact that you could do IO { cancel }, and therefore it could be done by user is correct. I more-less see consistency here, and I am ok that IO will be entering cancelled state before that `cancel` IO will get evaluated. That contract will IMHO not complicate implementation and will be consistent with Effect#runAsync. 

Also giving the results in `F` may easier compose with other code, such as Refs, Promises etc (i.e. cancel when given Promise completes). 

Note that even fact that we are in Effect.runAsync requiring IO instead `F` is complicating life a lot, as you need LiftIO to be defined for `F`, in case you need to perform some sequenced notifications (i.e. set a SyncRef). 

Simply want to avoid that unnecessary dependency with `cancel`. 

I think good example is the `cancel` in fs2.Promise, that has cancelation defined as F[Unit] and this combines and works pretty well, as @SystemFw pointed earlier.

**Comment by DavidGregory084** [comment]
@alexandru I also find the `() => Unit` signature a little weird (e.g. here https://github.com/DavidGregory084/pool/blob/master/src/main/scala/pool/Pool.scala#L379), resulting in some less than ideal code if cancellation involves any asynchronous cleanup actions:
```
Cancelable(() => Await.result(pool.close.runAsync(scheduler), Duration.Inf))
```
Obviously the above probably shouldn't be `Duration.Inf`, but you get the idea...

**Comment by alexandru** [comment]
@pchlupacek without the contract change, then it wouldn't be an issue from an implementation perspective, however I think the semantics for the current signature are more clear.

Side-effects are not suspended for `Async.async` for example, that callback being injected means an expectation for it to be called in a side-effectful manner, which we tolerate because `Async.async` is FFI that we use to wrap side effectful code.

But anyway, if `IO[Unit]` is desired without a change of contract, then I'm not strongly against it.

---

@DavidGregory084 your sample is precisely why I don't want it, because it should really be ...

```scala
Cancelable(() => pool.close.runAsync(scheduler))
```

If you're blocking for the result means you have *ordering* requirements and ordering requirements on cancellation (besides the cancellation signal being sent) are always wrong.

**Comment by DavidGregory084** [comment]
Yes, I think you're right, but if the signature is IO[A] then there is no onus on dummies like me to ignore the impulse to make the types line up and figure that out - the implementor of the Task/IO type can chuck the results away for me 🙂

**Comment by pchlupacek** [comment]
@alexandru I am not really sure how you want me to use cancel:  () => Unit in program that abstract over `F` w/o need to have Sync[F] // Effect[F] available.

**Comment by pchlupacek** [comment]
also I think that current Async.async signature is not ideal, and would like to propose to have it `F` wrapped too.

**Comment by alexandru** [comment]
@pchlupacek lets be more specific; can you give an example for which you'd need `Effect[F]` in `Async.async`?

**Comment by alexandru** [comment]
If that is indeed the case, then I can agree to `IO[Unit]`.

**Comment by SystemFw** [comment]
:+1: to wrapping `async` too.
If you change the signature and the code you're wrapping is side effectful, then all you need to do is `delay`.
But if you don't and the code you're wrapping is in IO, now you need to `unsafeRunSync`, or even `runAsync.unsafeRunSync`

**Comment by SystemFw** [comment]
@alexandru 
https://github.com/functional-streams-for-scala/fs2/blob/series/0.10/core/shared/src/main/scala/fs2/async/Promise.scala#L110

`go` is in `F`, I had to `unsafeRunAsync` it to make it fit into `async`

**Comment by SystemFw** [comment]
Another reason is the inconsistency with `runAsync`, where we do wrap in `F`

**Comment by pchlupacek** [comment]
@alexandru, lets assume you have somethin that produces callback after some `f` evaluates, i.e. 
```
Async.async { cb =>
  generateTime.flatMap { time =>
  checkState.get.ftatMap { s =>
    cb(time,s)
  }}
}

```
I hope it gives the idea ....

**Comment by pchlupacek** [comment]
@alexandru, @SystemFW, in fact I would prefer to have it `F` wrapped instead of `IO`. Only think I am sort of thinking why IO was chosen for Effect.runAsync is stack safety? I am not sure if that has to be part of Effect contract, as it requires to define `LiftIO` for `F`.

**Comment by SystemFw** [comment]
for `runAsync`, `IO` is necessary.
`runAsync` defines how to run an `F`, which is through a translation into a concrete type: IO. (otherwise you'd have to include a side effectful operation in the typeclasses, which I'm against).

For `async`, `F` should be fine.

**Comment by alexandru** [comment]
@SystemFw your sample is biased by your usage with `Promise`, which would need a redesign if `IO` is cancellable anyway — however the big problem is that you end up requiring `Effect` in actual normal usage:

```scala
def delayed(timespan: FiniteDuration)(implicit F : Async[F], sc: Scheduler) =
  F.async[Unit] { cb =>
    sc.scheduleOnce(timespan) {
      // Ooops, we'd need `Effect`
      cb(Right(()))
    }
  }
```

Also, this is not the purpose of this PR, would like to keep the conversation focused on `IO` — we can talk about improvements on the current type classes in another issue.

---

@pchlupacek you make a good case for `IO[Unit]`. Will see how this fits and get back.

**Comment by SystemFw** [comment]
> Also, this is not the purpose of this PR, would like to keep the conversation focused on IO — we can talk about improvements on the current type classes in another issue.

Agreed, apologies

**Comment by pchlupacek** [comment]
@SystemFw, @alexandru as side note, if we would have F[Unit] in `cb`, we can downgrade requirement for promise from (Effect, Ec) => (Sync, Ec), Perhaps even to `Sync only` imho.... (if I am reading correctly) and that would be just perfect...

**Comment by alexandru** [comment]
@SystemFw @pchlupacek I have opened an issue at https://github.com/typelevel/cats-effect/issues/123 — changing `Async` is not possible IMO, I described the reasons why and if there's anything you'd like to add, then do so in that issue.

**Comment by alexandru** [comment]
@SystemFw @pchlupacek @DavidGregory084 tried making the change in `IO.cancelable`  from `() => Unit` to `IO[Unit]` for that cancellation callback and realized that I don't want to do it. Reasons:

1. it's dishonest in regards to back-pressuring, which will not happen here; a `Unit` result on the other hand signals synchronous execution as a requirement
2. `IO` values should in general be referentially transparent, but `Cancelable` actions need to be in fact [idempotent](https://en.wikipedia.org/wiki/Idempotence)

This is why Monix defines an impure [Cancelable](https://monix.io/docs/2x/execution/cancelable.html) interface and uses that, because `.cancel()` has idempotence semantics and calling it multiple times should be safe.

As an example, here's one way of reading a file:

```scala
def readFile(file: File)(implicit ec: ExecutionContext): IO[String] =
  IO.cancelable { cb =>
    var isCancelled = false
    val in = new BufferedReader(
      new InputStreamReader(new FileInputStream(file), "utf-8"))

    ec.execute { () => 
      try {
        val buffer = new StringBuilder()
        var continue = true

        while (continue) {
          in.synchronize {
            if (isCancelled) return
            in.readLine() match {
              case null => 
                continue = false
              case line =>
                buffer.append(line)
            }          
          }
        }
        cb(Right(buffer.toString))
      } catch {
        case NonFatal(e) => cb(Left(e))
      }
    }

    // Cancellation logic
    () => in.synchronize {
      if (!isCancelled) {
        isCancelled = true
        // Needs idempotency ;-)
        in.close()
      }
    }
  }
```

So that cancellation reference will close over variables and cannot be referentially transparent in this context. Of course, there are cases in which this is OK. And also, in the case of the `IO.cancelable` builder in particular the implementation itself also has protections for idempotency (e.g. the function returned by `unsafeRunCancelable` is idempotent). But an `IO[Unit]` there sends the wrong message regardless.

@pchlupacek 

> I think good example is the cancel in fs2.Promise, that has cancelation defined as `F[Unit]` and this combines and works pretty well, as @SystemFw pointed earlier.

Cancellation is also defined as `F[Unit]` in the `Fiber` interface, being `F[Unit]` when you're speaking of it as a pure action.

But in the context of a signature such as `F[(F[A], F[Unit])]`, signatures which you've been using because `IO` is not cancellable, I heard that `F[Unit]` used there as a cancellation action has been problematic in FS2 precisely because of how referential transparency interacts with the idempotency requirement of cancellation actions.

---

My proposal to you is to let this signature be — and you try and make an exercise of redesigning `Promise` with this new `IO` in mind. Because it is my hunch right now that your `Promise` is designed around problems that won't exist anymore.

**Comment by SystemFw** [comment]
Mmm I'm unconvinced. 
I think the cancellation action should be in `F`, and that it should be up to IO to make sure that calling `cancel` is idempotent. I feel like it's `cancel` that needs to be idempotent, not the action you pass to it. 

>  I heard that F[Unit] used there as a cancellation action has been problematic in FS2 precisely because of how referential transparency interacts with the idempotency requirement of cancellation actions.

The problem there was calling `cancel` multiple times, but binding the inner `F[A]` multiple times or calling `cancel` before `F[A]`. Ideally the logic we have there should be handled by `IO` directly.

**Comment by SystemFw** [comment]
I also don't see why `() => Unit` and `F[Unit]` are any different in requiring or not requiring synchronous  execution. They don't seem to say anything about it (one way or another) to me.

**Comment by pchlupacek** [comment]
@alexandru I am completely lost how the "backpressure" is related to IO/cancellation itself. Can you please explain to me what do you mean by backpressure?

**Comment by alexandru** [comment]
@SystemFw @pchlupacek 

It's hard to explain more than what I already did. I'm not happy about a `() => Unit` either, due to not signaling precisely the semantics we need, but then we can't really put an impure `Cancelable` in `cats-effect`.

Therefore I realized this is not a hill I'm willing to die on. So I changed the signature:

```scala
def cancelable[A](k: (Either[Throwable, A] => Unit) => IO[Unit]): IO[A]
```

Is this OK with you?

**Comment by pchlupacek** [comment]
@alexandru thanks a lot for patience. Yeah that signature is ok with me.

**Comment by pchlupacek** [comment]
@alexandru wait, just to understand this completely. To build the IO, that is cancellable, I need to use IO.cancellable, and then how I precisely cancel this ? I think we have sort of dicussed that cancel shall return IO, instead of pure function, but this is just a constructor? I am also puzzled with api cancellable/uncancellable. 

Could you perhaps sum it up now ?  

Thank you.

**Comment by alexandru** [comment]
@pchlupacek 

> To build the IO, that is cancellable, I need to use IO.cancellable, and then how I precisely cancel this ?

As a side-effectful action you get `unsafeRunCancellable`, which gives you a `() => Unit` back that can be cancelled. This is a low-level API for `IO` specifically and won't get exposed in type classes.

Usually cancellation will happen under the hood on race conditions. For example the implementation of `parMap2` was changed in this PR to cancel the active task in case one of them finishes in error. 

At some point we'll also have an `IO.race` that will cancel the loser. Based on that you could then express a `timeout` operation that cancels the source in case a certain timespan was exceeded.
All of these are race conditions for which cancellation happens automatically.

Now there are instances in which an `IO` value gets exposed that's attached to a running process. For example this would be the result of a `start` operation, which in FS2 currently produces an `IO[IO[A]]`. These are `IO` values that are running concurrently and that the user might wish to cancel manually, therefore an operation like `start` is producing an `IO[Fiber[A]]`, where `Fiber` has a `cancel: IO[Unit]` on it.

I think the best thing you can do right now is to download this branch and play with it.

> I am also puzzled with api cancellable/uncancelable.

What exactly is confusing?

The `uncancelable` operation does what it says. Consider this very simple analogy:

```scala
type Run = () => (() => Unit)

def buildCancelable(): Run = {
  ???
}

def uncancelable(run: Run): Run = {
  () => {
    // execution
    run()
    // make cancellation a no-op
   () => {}
  }
}
```

**Comment by alexandru** [comment]
**NOTE:** I'm leaving this PR active until the end of this month.

In absence of any hard objections, it will get merged, then work can proceed for the type class hierarchy.

The plan I'm thinking of is to not do a new release unless we reach agreement on what happens to the current type class hierarchy.

**Comment by pchlupacek** [comment]
@alexandru ok lets go with this further. I still don't understand why we need `uncancellable`, but likely it won't hurt. Lets see how this fits in typeclases and then we could see. 

Thanks for explanation.

**Comment by alexandru** [comment]
@pchlupacek `uncancelable` is useful in cases where you really need to ensure that an `IO` has an atomic execution (either all of it, or none of it), effectively undoing any cancellation logic, needed sometimes as to not leave the program in an inconsistent state. This tends to happen when dealing with concurrent stuff a lot.

For example with Monix or with FS2 streams, the user could specify logic to be executed at the end of the stream, via the stream's own finalization protocol. And then when you end up doing something like:

```scala
// Pseudocode
val result: IO[A] = stream.fold(???)
```

Well, there will be cases in which the user will not know whether this resulting `IO` can be safely cancelled or not. It will also depend on the stream's own implementation, although making cancelation automatic is always problematic, because its logic is going to be concurrent, versus a stream's own interruption mechanism that's properly ordered (well I can speak more for Monix's `Iterant`, but I'm fairly sure this applies to FS2's `Stream` as well).

Of course an `IO.bracket` operation will be available too, but that won't be sufficient, this being IMO a weakness in Haskell's `IO`, because `bracket` forces you to design solutions around it, it forces you to deal with resource handling within the context of `IO`, which isn't very compatible with streaming approaches that are doing higher-level resource handling, like FS2's `Stream` or Monix's `Iterant`.

So `uncancelable` can ensure that an `IO` value has an atomic execution and in that regard it never hurts for it to be there.

**Comment by pchlupacek** [comment]
@alexandru ok got it. Thanks.

**Comment by durban** [comment]
This program:
```scala
object CancellableIOExample {

  val promise = Promise[Int]()

  val task: IO[Unit] = IO.cancelable[Int] { cb =>
    println("starting")
    promise.future.foreach { i =>
      println("calling callback")
      cb(Right(i))
    }
    IO { println("cancelling") }
  }.flatMap { i =>
    IO { println(s"completed with $i") }
  }

  def main(args: Array[String]): Unit = {
    val fib = task.start.unsafeRunSync()
    Thread.sleep(100)
    fib.cancel.unsafeRunSync()
    Thread.sleep(100)
    promise.success(42)
    Thread.sleep(100)
    println("end")
  }
}
```
prints this:
```
starting
cancelling
calling callback
completed with 42
end
```

That is, calling the callback takes effect even after cancellation. Is this intentional?

**Comment by alexandru** [comment]
@durban 

> That is, calling the callback takes effect even after cancellation. Is this intentional?

Yes. In regards to cancellation, nothing happens automatically that the user doesn't intend, cancellation being a signal that the producer must notice and act on it, explicitly:

```scala
// Version one, which triggers error on cancel:
IO.cancelable[Int] { cb =>
  println("starting")
  promise.future.foreach { i =>
    println("calling callback")
    cb(Right(i))
  }
  IO { 
    promise.tryFailure(new CancellationException)
  }
}

// Version two, which makes it non-terminating on cancel:
IO.cancelable[Int] { cb =>
  val active = new AtomicBoolean(true)
  println("starting")
  promise.future.foreach { i =>
    if (active.getAndSet(false)) {
      println("calling callback")
      cb(Right(i))
    }
  }
  IO { 
    if (active.getAndSet(false))
      println("cancelling")
  }
}
```

**Comment by alexandru** [comment]
If nobody will mind, I'd like to merge this on Monday because I have some time off and would like to focus on https://github.com/typelevel/cats-effect/pull/126

No release will be made, so we can still backtrack if needed.

**Pull request merged**